trino-lb/src/http_server/admin/mod.rs

```diff
@@ -0,0 +1,170 @@
+) -> Result<Json<ClusterActivationResponse>, Error> {
+    state
+        .metrics
+        .http_counter
+        .add(1, &[KeyValue::new("resource", "post_activate_cluster")]);
+
+    set_cluster_activation(state, basic_auth, &cluster_name, true).await
+}
+
+/// Deactivate a Trino Cluster, so that it doesn't receive any new queries.
+///
+/// This is useful for maintenance actions (in combination with activation).
+#[instrument(name = "POST /admin/deactivate-cluster/{cluster_name}", skip(state))]
+pub async fn post_deactivate_cluster(
+    TypedHeader(Authorization(basic_auth)): TypedHeader<Authorization<Basic>>,
+    State(state): State<Arc<AppState>>,
+    Path(cluster_name): Path<TrinoClusterName>,
+) -> Result<Json<ClusterActivationResponse>, Error> {
+    state
+        .metrics
+        .http_counter
+        .add(1, &[KeyValue::new("resource", "post_deactivate_cluster")]);
+
+    set_cluster_activation(state, basic_auth, &cluster_name, false).await
+}
+
+/// Get the status of the Trino clusters
+#[instrument(name = "GET /admin/cluster-status", skip(state))]
+pub async fn get_cluster_status(
+    State(state): State<Arc<AppState>>,
+) -> Result<Json<BTreeMap<TrinoClusterName, ClusterStats>>, Error> {
+    state
+        .metrics
+        .http_counter
+        .add(1, &[KeyValue::new("resource", "get_cluster_status")]);
+
+    let cluster_stats = state
+        .cluster_group_manager
+        .get_all_cluster_stats()
+        .await
+        .context(GetAllClusterStatesSnafu)?;
+
+    Ok(Json(
+        cluster_stats
+            .into_iter()
+            .map(|(cluster, stats)| (cluster.name.clone(), stats))
+            .collect(),
+    ))
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ClusterActivationResponse {
+    state: ClusterState,
+}
+
+#[instrument(skip(state))]
+async fn set_cluster_activation(
+    state: Arc<AppState>,
+    basic_auth: Basic,
+    cluster_name: &TrinoClusterName,
+    activation: bool,
+) -> Result<Json<ClusterActivationResponse>, Error> {
+    match &state.config.trino_lb.admin_authentication {
+        Some(TrinoLbAdminAuthenticationConfig::BasicAuth { username, password }) => {
+            ensure!(
+                basic_auth.username() == username && basic_auth.password() == password,
+                InvalidAdminCredentialsSnafu {}
+            );
+        }
+        None => return Err(Error::NoAdminAuthenticationMethodDefined),
+    }
+
+    ensure!(
+        state.config.cluster_in_config(cluster_name),
+        UnknownClusterSnafu {
+            cluster: cluster_name
+        }
+    );
+
+    let desired_state = if activation {
+        info!(cluster = cluster_name, "Re-activating Trino cluster");
+        ClusterState::Unknown
+    } else {
+        info!(cluster = cluster_name, "Deactivating Trino cluster");
+        ClusterState::Deactivated
+    };
+
+    state
+        .persistence
+        .set_cluster_state(cluster_name, desired_state.clone())
```

Comment:
note: I think the `ClusterState` enum can derive `Copy` which would remove the need to clone here. This won't have any performance gains, but it removes the `.clone()` call.

Reply:
Makes sense, bf71769